### ISSUE TITLE
Further tests of overheads in new SLURM backend

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal-server#615]
With the [new stress-test tasks](https://github.com/fractal-analytics-platform/fractal-demos/issues/40#issuecomment-1504787187), we can now require a certain amount of both memory and running time for a given task. Let's use this to better understand overheads (ref #597).

Let's prepare a workflow which runs on 400 components, and where each parallel tasks needs a certain amount of time (e.g. 1 second). Then we can explore different scenarios, like:
* 20 jobs with 20 tasks each
* 100 jobs with 4 tasks each
* 400 jobs with 1 task each

The goal of this analysis is twofold:
* Making sure that one such workflow can run in a reasonably small amount time (this is to be defined, but it should be closer to 10 minutes than to 2 hours..);
* Better understanding the overheads dependence on the number of jobs.

**Comment by tcompa** [comment]
A first test (this is for the record, but analysis is in progress).

I set up execution of a single task (taking 1 second per execution), on 200 parallel components.
Batching configuration splits the 200 tasks into 13 SLURM jobs of (up to) 16 tasks each.

Workflow log shows a total duration of approx 8 minutes (note: this was a single task), which are essentially all made of overhead.

I will report on this further.

**Comment by tcompa** [comment]
One more data point..

* 400 components
* Tasks that takes 0.1 seconds each
* 17 SLURM jobs with (up to) 16 tasks each

Workflow logs
```
2023-04-13 08:52:24,049; INFO; START workflow "Workflow 50"
2023-04-13 08:52:24,404; INFO; SUBMIT 0-th task (name="prepare_metadata")
2023-04-13 08:52:55,591; INFO; END    0-th task (name="prepare_metadata")
2023-04-13 08:52:55,593; INFO; SUBMIT 1-th task (name="sleep_task")
2023-04-13 09:08:07,290; INFO; END    1-th task (name="sleep_task")
2023-04-13 09:08:08,291; INFO; END workflow "Workflow 50"
```

Total duration: ~15 minutes. Keep in mind that this is _per-task_, and a workflow with N parallel tasks (say N=5) may take N times longer.

Task logs (first start, and last end)
```
$ cat 1_par_*.err | sort | head -n1
2023-04-13 08:52:58,228; INFO; START sleep_task task

$ cat 1_par_*.err | sort | tail -n1
2023-04-13 09:06:56,382; INFO; EXITING sleep_task
```

Note that the slow-`sudo` overhead is not taking place fully in a single time interval at the end of the 400 executions, but rather in the middle. This is clear from the fact that the last task that ends does it at 09:06:56, and the workflow log sees that the tasks are all over at 09:08:08 (~1 minute later). This means that most of the overhead (which has a total size of around 15 minutes) already took place earlier.

I will debug a bit further - and also report about different batching scenarios.

## Decision to be taken
(@jluethi @gusqgm)

The current overhead is something around 15 minutes for a single parallel task running over 400 components (this is also with a `sudo` time of 0.35 seconds, and with a certain (reasonable) batching scheme).

**Is this  acceptable?**

My first guess is "probably not", since a 5-parallel-tasks workflow would take more than one hour.

If so, we should further optimize this procedure - with the obvious candidate being the current per-file procedure (read with `subprocess.run("sudo -u user cat file1"`, and then write from python), which we could maybe replace with a single `cp` command (to be checked: would it have the right permissions both on read and write paths?)

**Comment by tcompa** [comment]
Same as in the last comment, but with 400 single-task SLURM jobs.

```
2023-04-13 09:30:08,542; INFO; START workflow "Workflow 52"
2023-04-13 09:30:08,911; INFO; SUBMIT 0-th task (name="prepare_metadata")
2023-04-13 09:30:39,995; INFO; END    0-th task (name="prepare_metadata")
2023-04-13 09:30:39,997; INFO; SUBMIT 1-th task (name="sleep_task")
2023-04-13 09:57:28,072; INFO; END    1-th task (name="sleep_task")
2023-04-13 09:57:29,073; INFO; END workflow "Workflow 52
```

Total runtime: ~27 minutes
-> different batching has a relevant effect (although its source is not necessarily fractal-related, as I think that the number of `sudo` calls is identical.. to be checked)

Tasks
```
$ cat 1_p*err | sort | head -n1
2023-04-13 09:30:42,370; INFO; START sleep_task task

$ cat 1_p*err | sort | tail -n1
2023-04-13 09:57:19,242; INFO; EXITING sleep_task
```

---

Summary
```
17 jobs with (up to) 16 tasks each
total runtime: ~15 minutes

400 jobs with 1 task each
total runtime: ~27
```

**Comment by jluethi** [comment]
I will try running this on FMI servers as well (just the version with batched jobs). It would be good to understand how much of this is due to slow sudo and how much we'd have on any reasonable slurm backend.

> **Is this acceptable?**
> 
> My first guess is "probably not", since a 5-parallel-tasks workflow would take more than one hour.

It depends. It certainly isn't optimal. If it's "only" due to slow sudo, I'd argue for trying to address this first, instead of reworking out approach here. If it has high level of overhead at FMI as well, then we certainly need to address this.

---

```
17 jobs with (up to) 16 tasks each
total runtime: ~15 minutes

400 jobs with 1 task each
total runtime: ~27
```

This is indeed surprising, shows the usefulness of job batching I guess

**Comment by tcompa** [comment]
> If it's "only" due to slow sudo, I'd argue for trying to address this first

The first attempt (resizing the fileserver to more CPUs and memory) didn't work (although it will possibly  mitigate other cases of slow CLI at UZH). We'll have to wait and see if more detailed debugging leads to a possible solution.

**Comment by tcompa** [comment]
Another information, for the record, is that running a 400-components task (with each task taking 0.01 seconds) on a local backend takes something like 6 seconds.

This is just meant to be a sanity check: the functions in `runner/common.py` and `runner/_common.py` are not slow in general, but only when executed through the SLURM backend.

**Comment by jluethi** [comment]
I ran the stress test on the FMI slurm testing server (not the main cluster, but I think things should be very similar there).

It ran for 400 components, "tasks_per_job": 20, "parallel_tasks_per_job": 20 => 20 total slurm jobs. I kept the sleep time at 1 (=> 1s?)
It finished actually running those slurm jobs within a few seconds for each (slurm jobs ran for about 3-4s each). It could only run 2 jobs at a time given the resources of that machine. Within each slurm job, there was basically no waiting for anything. After it finished running all the jobs, it took a while (~2min) for Fractal server to realize that they were all finished (=> the sudo copying overhead I assume).

```
2023-04-14 12:21:01,705; INFO; START workflow "Workflow 4"
2023-04-14 12:21:01,757; INFO; SUBMIT 0-th task (name="prepare_metadata")
2023-04-14 12:21:04,188; INFO; END    0-th task (name="prepare_metadata")
2023-04-14 12:21:04,191; INFO; SUBMIT 1-th task (name="sleep_task")
2023-04-14 12:23:16,761; INFO; END    1-th task (name="sleep_task")
2023-04-14 12:23:17,763; INFO; END workflow "Workflow 4"
```

I recorded the time when the last slurm job finished: 12:21:48. Thus, running the 20 jobs (2 at a time) took 44s. After that, 1min 30s were spent waiting for the copying overhead, for the server to realize the jobs were done.

I ran the example again with 2 parallel tasks after each other, both with those exact same settings. Here are the workflow logs for this:

```
2023-04-14 12:24:31,896; INFO; START workflow "Workflow 5"
2023-04-14 12:24:31,951; INFO; SUBMIT 0-th task (name="prepare_metadata")
2023-04-14 12:24:34,371; INFO; END    0-th task (name="prepare_metadata")
2023-04-14 12:24:34,377; INFO; SUBMIT 1-th task (name="sleep_task")
2023-04-14 12:26:45,828; INFO; END    1-th task (name="sleep_task")
2023-04-14 12:26:45,832; INFO; SUBMIT 2-th task (name="sleep_task")
2023-04-14 12:29:01,108; INFO; END    2-th task (name="sleep_task")
2023-04-14 12:29:02,111; INFO; END workflow "Workflow 5"
```

The last slurm job finished at 12:27:28, thus again about 1.5 min until Fractal server logs the task as END. Overall, the performance was very similar, it seems we reproducibly have a ~1.5 min overhead for the 20 jobs à 20 runs (=> 400 components, batched) at the FMI.

---

In conclusion, the slurm copying overhead is also present in the FMI scenario, but it is much smaller. Having to wait 1.5 minutes after each task over 400 components completes is not optimal, but not something we'd need to urgently address. For a 5 parallel task workflow, that would be ~7.5 min overhead. I'd prefer not to have it, but it's no deal breaker.

**Comment by jluethi** [comment]
Ok, first real-life performance data on this: Running 1 FOV per well on a 384 well plate (not available in all wells => only 280 components to be processed) on the FMI cluster (upcoming user example 08_HCA_plate):

This ran within 1h 10min. And it only ran max. half of the job at once due to the cluster being busy at the time. The heuristics were working well.

Here is the workflow log:
```
2023-04-14 14:12:39,458; INFO; START workflow "Workflow HCA_Plate_mainCluster_F016-4"
2023-04-14 14:12:39,629; INFO; SUBMIT 0-th task (name="Create OME-Zarr structure")
2023-04-14 14:24:20,525; INFO; END    0-th task (name="Create OME-Zarr structure")
2023-04-14 14:24:20,529; INFO; SUBMIT 1-th task (name="Convert Yokogawa to OME-Zarr")
2023-04-14 14:43:27,439; INFO; END    1-th task (name="Convert Yokogawa to OME-Zarr")
2023-04-14 14:43:27,442; INFO; SUBMIT 2-th task (name="Copy OME-Zarr structure")
2023-04-14 14:46:44,368; INFO; END    2-th task (name="Copy OME-Zarr structure")
2023-04-14 14:46:44,371; INFO; SUBMIT 3-th task (name="Maximum Intensity Projection")
2023-04-14 15:01:32,162; INFO; END    3-th task (name="Maximum Intensity Projection")
2023-04-14 15:01:32,171; INFO; SUBMIT 4-th task (name="Cellpose Segmentation")
2023-04-14 15:12:23,586; INFO; END    4-th task (name="Cellpose Segmentation")
2023-04-14 15:12:23,592; INFO; SUBMIT 5-th task (name="Napari workflows wrapper")
2023-04-14 15:22:59,080; INFO; END    5-th task (name="Napari workflows wrapper")
2023-04-14 15:23:00,084; INFO; END workflow "Workflow HCA_Plate_mainCluster_F016-4"
```

Every task took 10-20 minutes to run. Thus, for the case of testing on 1 FOV per well for 280 wells, we have somewhere in the order of 15% overhead (where slurm is not running anything for us).

I'll run tests with the full well next, let's see how they perform :)

**Comment by jluethi** [comment]
Well, I have some mind-blowing results to report here from the FMI cluster. I ran the whole HCA plate, 379 wells, each consisting of ~15-25 FOV, 16 Z stacks, 4 Channels. In total, compressed it's about 1.7 TB of image data. I optimized the parameters a bit so that Fractal could use more of the nodes on the cluster and the cluster was less busy than earlier in the afternoon. Thus, for each task, it could run 10 jobs in parallel, each job had up to 40 tasks being processed at once (while before, only ~half the jobs ran at once).

It ended up processing the whole HCA plate, the whole 1.7 TB of data in... 1h 7 min!!!!
It was faster at processing EVERYTHING than processing the subset.

Here is the workflow log:
```
2023-04-14 16:38:13,644; INFO; START workflow "Workflow HCA_Plate_Take2_D8_R1-1"
2023-04-14 16:38:13,828; INFO; SUBMIT 0-th task (name="Create OME-Zarr structure")
2023-04-14 16:49:34,100; INFO; END    0-th task (name="Create OME-Zarr structure")
2023-04-14 16:49:34,105; INFO; SUBMIT 1-th task (name="Convert Yokogawa to OME-Zarr")
2023-04-14 17:12:38,507; INFO; END    1-th task (name="Convert Yokogawa to OME-Zarr")
2023-04-14 17:12:38,512; INFO; SUBMIT 2-th task (name="Copy OME-Zarr structure")
2023-04-14 17:16:44,989; INFO; END    2-th task (name="Copy OME-Zarr structure")
2023-04-14 17:16:44,995; INFO; SUBMIT 3-th task (name="Maximum Intensity Projection")
2023-04-14 17:28:59,614; INFO; END    3-th task (name="Maximum Intensity Projection")
2023-04-14 17:28:59,621; INFO; SUBMIT 4-th task (name="Cellpose Segmentation")
2023-04-14 17:40:00,277; INFO; END    4-th task (name="Cellpose Segmentation")
2023-04-14 17:40:00,281; INFO; SUBMIT 5-th task (name="Napari workflows wrapper")
2023-04-14 17:45:24,189; INFO; END    5-th task (name="Napari workflows wrapper")
2023-04-14 17:45:25,190; INFO; END workflow "Workflow HCA_Plate_Take2_D8_R1-1"
```

I checked the images & the logs, it really processed the whole plate. I do not understand how that was possible so fast! I was expecting 12-16h or more (see above)!!!

The thing that made it faster than the single FOV was that it was running more things in parallel (about a factor of 2). Which means that in the same conditions, it appears to take about 2 times as long to process 15x the amount of data. Which is absolutely nuts! I can't believe the GPFS filesystem & the whole infrastructure is scaling to that degree!

If we can process 1.7 TB in under 2h, I don't think we'd need to optimize speeds ;) Now of course, there is processing that could be done that would take significantly more time. Doing more processing in 3D (e.g. 3D segmentation, 3D registration) would require significantly more actual compute time. But it's still incredible that the infrastructure scales to this degree!

Of course, it also tells us that the bottleneck in the processing of a single FOV clearly isn't the CPU/GPU side and, on FMI infrastructure, apparently also not the IO speed of reading & writing these amounts of data. So there's likely potential to make the subset processing even faster and decrease overhead.

Anyhow, I'm still in this mood & signing off for the weekend:
![C8qlRE6VYAAWdyt](https://user-images.githubusercontent.com/18033446/232100419-ad73784c-b799-4101-b2df-4f19b64946f9.jpg)

**Comment by tcompa** [comment]
One more example at UZH, after switching to a separate fileserver.

The new sudo time is about 10-15 ms (to be compared with ~350 ms, in the previous situation).

We ran a workflow with a single very short task (0.1 seconds), parallelized over 400 wells (with 16 tasks per job), and the total runtime was1:30 minutes. To be compared with 15 minutes, as in https://github.com/fractal-analytics-platform/fractal-server/issues/615#issuecomment-1506644417.

```
2023-04-21 09:15:15,447 - WF1_job1 - INFO - Start execution of workflow "Workflow 11"; more logs at /net/nfs4/pelkmanslab-fileserver-fractal/data/active/fractal/deployments/server_01/artifacts/proj_0000001_wf_0000001_job_0000001/workflow.log
2023-04-21 09:15:15,447 - WF1_job1 - DEBUG - fractal_server.__VERSION__: 1.2.1
2023-04-21 09:15:15,447 - WF1_job1 - DEBUG - FRACTAL_RUNNER_BACKEND: slurm
2023-04-21 09:15:15,447 - WF1_job1 - DEBUG - slurm_user: test01
2023-04-21 09:15:15,447 - WF1_job1 - DEBUG - worker_init: None
2023-04-21 09:15:15,447 - WF1_job1 - DEBUG - input metadata: {}
2023-04-21 09:15:15,447 - WF1_job1 - DEBUG - input_paths: ['/data/homes/test01/fractal-demos/examples/99_stress_test/tmp']
2023-04-21 09:15:15,447 - WF1_job1 - DEBUG - output_path: /data/homes/test01/fractal-demos/examples/99_stress_test/output-11
2023-04-21 09:15:15,463 - WF1_job1 - DEBUG - job.id: 1
2023-04-21 09:15:15,463 - WF1_job1 - DEBUG - job.working_dir: /net/nfs4/pelkmanslab-fileserver-fractal/data/active/fractal/deployments/server_01/artifacts/proj_0000001_wf_0000001_job_0000001
2023-04-21 09:15:15,463 - WF1_job1 - DEBUG - job.workflow_dir_user: /net/nfs4/pelkmanslab-fileserver-test01/data/homes/test01/fractal-demos/examples/cache/20230421_091515_proj_0000001_wf_0000001_job_0000001
2023-04-21 09:15:15,463 - WF1_job1 - DEBUG - START workflow "Workflow 11"
2023-04-21 09:15:15,480 - WF1_job1 - DEBUG - SUBMIT 0-th task (name="prepare_metadata")
2023-04-21 09:15:17,717 - WF1_job1 - DEBUG - END    0-th task (name="prepare_metadata")
2023-04-21 09:15:17,718 - WF1_job1 - DEBUG - SUBMIT 1-th task (name="sleep_task")
2023-04-21 09:16:44,429 - WF1_job1 - DEBUG - END    1-th task (name="sleep_task")
2023-04-21 09:16:45,431 - WF1_job1 - DEBUG - END workflow "Workflow 11"
```

I don't have much to add: this can still be improved, of course, but it starts to be close to irrelevant for anything apart from tiny tasks.

**Comment by jluethi** [comment]
This is awesome to see, glad the new fileserver makes such a difference! Very different in timings to the FMI setup then :)

And I agree, no need to focus on optimizing this further for the time being.

If we come back to this, my current conclusions on the initial questions are:

> 1. Making sure that one such workflow can run in a reasonably small amount time (this is to be defined, but it should be closer to 10 minutes than to 2 hours..);

As long as the infrastructure works well (e.g. FMI filesystems or new UZH fractal share), that looks fine. 

> 2. Better understanding the overheads dependence on the number of jobs.

Overhead seems to scale with number of components, but job batching has a surprisingly big influence on reducing the overhead.

(@tcompa feel free to reopen if there's more we should look into here)
